File: BuilderDesignPattern/SQLQuery_Example/Products/SQLQuery.java
package BuilderDesignPattern.SQLQuery_Example.Products;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SQLQuery {
	private final String table;
	private final List<String> columns;
	private final List<String> conditions;
	private final String orderBy;
	private final String orderDirection;
	private final int limit;
	private final int offset;

	public SQLQuery(Builder builder) {
		this.table = builder.table;
		this.columns = List.copyOf(builder.columns);
		this.conditions = List.copyOf(builder.conditions);
		this.orderBy = builder.orderBy;
		this.orderDirection = builder.orderDirection;
		this.limit = builder.limit;
		this.offset = builder.offset;
	}

	public static class Builder {
		private final String table;
		private List<String> columns = new ArrayList<String>();
		private List<String> conditions = new ArrayList<String>();
		private String orderBy;
		private String orderDirection = "ASC";
		private int limit;
		private int offset;

		public Builder(String table) {
			this.table = table;
		}

		public Builder select(String... columns) {
			this.columns.addAll(Arrays.asList(columns));
			return this;
		}

		public Builder where(String condition) {
			this.conditions.add(condition);
			return this;
		}

		public Builder orderBy(String column, String direction) {
			this.orderBy = column;
			this.orderDirection = direction;
			return this;
		}

		public Builder limit(int limit) {
			this.limit = limit;
			return this;
		}

		public Builder offset(int offset) {
			this.offset = offset;
			return this;
		}

		public SQLQuery build() {
			return new SQLQuery(this);
		}
	}

	public String toSQL() {
		StringBuilder query = new StringBuilder("SELECT ");
		query.append(columns.isEmpty() ? "*" : String.join(", ", columns));
		query.append(" FROM ").append(table);
		if (!conditions.isEmpty()) {
			query.append(" WHERE ").append(String.join(" AND ", conditions));
		}
		if (orderBy != null) {
			query.append(" ORDER BY ").append(orderBy).append(" ").append(orderDirection);
		}
		if (limit > 0) {
			query.append(" LIMIT ").append(limit);
		}
		if (offset > 0) {
			query.append(" OFFSET ").append(offset);
		}
		query.append(";");
		return query.toString();
	}
}

File: BuilderDesignPattern/SQLQuery_Example_Python/Products/SQLQuery.py
from typing import Final, Optional

class SQLQuery:

	class Builder:

		def __init__(self, table: str):
			self._table: Final[str] = table
			self._columns: list[str] = []
			self._conditions: list[str] = []
			self._orderBy: str = ""
			self._orderDirection: str = "ASC"
			self._limit: Optional[int] = None
			self._offSet: Optional[int] = None

		def select(self, *columns: str) -> "SQLQuery.Builder":
			self._columns.extend(columns)
			return self

		def where(self, condition: str) -> "SQLQuery.Builder":
			self._conditions.append(condition)
			return self

		def orderBy(self, column: str, direction: str) -> "SQLQuery.Builder":
			self._orderBy = column
			self._orderDirection = direction
			return self

		def limit(self, count: int) -> "SQLQuery.Builder":
			self._limit = count
			return self

		def offSet(self, offSet: int) -> "SQLQuery.Builder":
			self._offSet = offSet
			return self

		def build(self) -> "SQLQuery":
			return SQLQuery(self)

	def __init__(self, builder: "Builder"):
		self.__table: Final[str] = builder._table
		self.__columns: Final[ list[str] ] = builder._columns
		self.__conditions: Final[ list[str] ] = builder._conditions
		self.__orderBy: Final[ str ] = builder._orderBy
		self.__orderDirection: Final[ str ] = builder._orderDirection
		self.__limit: Final[ int ] = builder._limit
		self.__offSet: Final[ int ] = builder._offSet

	def __str__(self) -> str:
		# 1. Start with SELECT and handle columns
		# Accessing name-mangled __columns and __table
		column_part = ", ".join(self.__columns) if self.__columns else "*"
		query_parts = [ f"SELECT {column_part} FROM {self.__table}" ]

		# 2. Add WHERE conditions (from self.__conditions list)
		if self.__conditions:
			query_parts.append(f"WHERE {' AND '.join(self.__conditions)}")

		# 3. Add ORDER BY (from self.__orderBy and self.__orderDirection)
		if self.__orderBy:
			# Defaulting to ASC if not provided, though your builder sets a default
			direction = self.__orderDirection if self.__orderDirection else "ASC"
			query_parts.append(f"ORDER BY {self.__orderBy} {direction}")

		# 4. Add LIMIT
		# Check if limit is not None and greater than 0
		if self.__limit is not None and self.__limit > 0:
			query_parts.append(f"LIMIT {self.__limit}")

		# 5. Add OFFSET
		if self.__offSet is not None and self.__offSet > 0:
			query_parts.append(f"OFFSET {self.__offSet}")

		# Join with spaces and terminate with a semicolon
		return " ".join(query_parts) + ";"
